Feature asg instance status wait

zollie opened the pull request:
I'm not sure this is the most elegant way to handle this so feedback is welcome ...

Let me give you the scenario:

I have a public and private VPC. A NAT instance is launched in an ASG in the public VPC with min and max set to 1.

Next instances are launched in an ASG in a private subnet. These require the NAT instance to be up, running, and initialized to properly initialize themselves.

An ASG will report instances Healthy and InService before they are "Initialized". This is too soon.

This PR adds a boolean "wait_for_instance_system_status_ok" that waits for an ASG instance's SystemStatus to be "ok" before considering it ready. That is after "Initializing" is done.

I should add, this is working well for me. 

If merged, I will update docs ...

zollie commented:
Bump. 

I'm confident this or something like it is needed in Terraform. It's the best for to ensure an ASG is both Healthy with it's instances "Initialed" before moving on to it's dependencies. 

I'm not married to the name of the boolean so open in that regard. Code critique is welcome as well.

phinze commented:
Hi @zollie - thanks for digging into this!

It's amazing how many different health / status / state fields come into play here.

phinze commented:
Hi @zollie - thanks for digging into this!

It's amazing how many different health / status / state fields come into play here. 

I believe that each resource should block until it can guarantee it's ready to perform service. With that in mind, I'm wondering if rather than a configurable if this should just be added to the default behavior, just like the ASG Health status poll.

I can't think of a scenario where anybody would legitimately want to skip it, and it would simplify the code change as well.

What do you think?

zollie commented:
Sounds good to me. Agreed, if one ASG depends on another than the desired capacity it is seeking should really include the "Initialization" phase for those instances as well. I will update and resubmit PR.

zollie commented:
So I just got back to this. I did some testing with this as the default behavior and the downside is the dependent ASG is the last thing to complete as it also is waiting for it's desired instances to initialize. But, there is nothing else dependent on those instances. Does that make sense? The `apply` is longer now. Instances can take a while to initialize depending, for me anyway, on the stuff going on in cloud-init.

zollie commented:
Updated PR to make waiting on EC@ instance status="ok" the default ...

phinze commented on CHANGELOG.md:
```diff
@@ -2,7 +2,17 @@
+
+IMPROVEMENTS:
+
+  * provider/aws: Create RDS databases from snapshots [GH-2062]
```
We leave off CHANGELOG updates on PRs to avoid conflicts - the maintainers will update this post-merge.

phinze commented on builtin/providers/aws/resource_aws_db_instance.go:
```diff
@@ -238,6 +238,19 @@ func resourceAwsDbInstance() *schema.Resource {
+				Type:     schema.TypeBool,
+				Computed: false,
+				Optional: true,
+			},
```
Looks like these diffs are happening because the branch needs a rebase.

phinze commented:
Looking good - I think the branch could use a rebase

zollie commented:
I think those diffs your referencing were rebased in earlier. I rebased a fresh master before last push and just did again to be sure, and there were no changes. I changed only 1 file: `resource_aws_autoscaling_group.go`

zollie commented:
Oh wait, I think I did screw something up. I see those diffs locally. Let me check.

zollie commented:
@phinze You were correct. I totally missed a conflict from the rebase. Should be good now.

phinze commented:
@zollie looks like there's still a diff in the Google provider.

Can you do a [rebase and squash](http://blog.steveklabnik.com/posts/2012-11-08-how-to-squash-commits-in-a-github-pull-request)?

(If the git acrobatics are getting to be too much for you - one of us can rebase/squash for you and leave you as an author on the commit.)

zollie commented:
I'm generally pretty good with git. That last rebase was a squash of 3 commits into 1. I was just having a brain fart previous to that.  

I am a little more confused by github though? I'm trying to figure out what you see as I just pulled `upstream/master` and diff'ed it with `f-asg-instance-status-wait` and I only see my changes to the 1 file now. 

Can you see this [diff](https://github.com/hashicorp/terraform/compare/master...zollie:f-asg-instance-status-wait)?

phinze commented:
Oh weird. I was just looking at https://github.com/hashicorp/terraform/pull/2617/files which has a massive diff. Perhaps it's just GitHub that is confused then.

Can you close this and open a fresh PR and see if that gets it to figure itself out?

zollie commented:
Yep, sounds good.

zollie commented:
FYI, I put this in PR 2742
